openwisp_radius/saml/backends.py

```diff
@@ -0,0 +1,28 @@
+from django.core.exceptions import ObjectDoesNotExist
+from djangosaml2.backends import Saml2Backend
+
+from .. import settings as app_settings
+
+
+class OpenwispRadiusSaml2Backend(Saml2Backend):
+    def _update_user(self, user, attributes, attribute_mapping, force_save=False):
+        if not app_settings.SAML_UPDATES_USERNAME and not user._state.adding:
+            # Skip updating user's username if the user didn't signed up
+            # with SAML registration method.
+            try:
+                attribute_mapping = attribute_mapping.copy()
+                if user.registered_user.method != 'saml':
+                    for key, value in attribute_mapping.items():
+                        if 'username' in value:
+                            break
+                    if len(value) == 1:
+                        attribute_mapping.pop(key, None)
+                    else:
+                        attribute_mapping[key] = []
+                        for attr in value:
+                            if attr != 'username':
+                                attribute_mapping[key].append(attr)
+
+            except ObjectDoesNotExist:
+                pass
+        return super()._update_user(user, attributes, attribute_mapping, force_save)
```

Comment:
I think something is in how this works.

I believe if the user is created with SPID, we shall always update the username no matter what.
If the user was not created with SPID, we shall not update their username unless the setting is set to true.
The docs should reflect this difference.

I'd rename the setting something like `SAML_UPDATES_PRE_EXISTING_USERNAME` to make it more explicit.

Reply:
> I believe if the user is created with SPID, we shall always update the username no matter what.
If the user was not created with SPID, we shall not update their username unless the setting is set to true.
The docs should reflect this difference.

The code functions like this in the current state. I have updated name of the setting.